Add 2.0-2.3 facts api compat (ansible_facts(), get_all_facts())

alikins opened the pull request:
These are intended to provide compatibilty for modules that
use 'ansible.module_utils.facts.ansible_facts' and
'ansible.module_utils.facts.get_all_facts' from 2.0-2.3 facts
API.

Fixes #25686

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/module_utils/facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (get_all_facts 32c64c7c45) last updated 2017/07/25 13:52:15 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```


##### ADDITIONAL INFORMATION

Not a huge fan of the code/boilerplate in facts/__init__.py, but not sure there is a better approach. (Could use python namespace packages maybe? That seems undesirable though...)

alikins committed:
Add 2.0-2.3 facts api compat (ansible_facts(), get_all_facts())

These are intended to provide compatibilty for modules that
use 'ansible.module_utils.facts.ansible_facts' and
'ansible.module_utils.facts.get_all_facts' from 2.0-2.3 facts
API.

Fixes #25686

alikins committed:
rm ansible.constants import from module_utils.facts.compat

No 'ansible.constants' when remote modules get invoked, so
that was never going to work.

Just use a hard coded default for gather_subset/gather_timeout
instead of trying to load it from non existent config if the
module params dont include it.

mtnbikenc commented:
@alikins I'm testing again against openshift-ansible and received a TypeError.
`TypeError: Bad subset 'facter' given to Ansible.`

Is `facter` being removed in 2.4?  It does not appear we are using any facter_ facts anyway.

alikins commented:
@mtnbikenc post the -vvv error message and playbook snippet?

afaik, 'facter' should be fine. It's still in 2.4 and in theory should work for compat api as well.

alikins committed:
include 'external' collectors in compat ansible_facts()

alikins commented:
@mtnbikenc think I see the problem, I wasnt including the 'external' collectors in the list of valid collectors in ansible_facts. (ie, 'facter' and 'ohai'). Just pushed an update that fixes that.

alikins committed:
Add facter/ohai back to the valid collector classes

Had unintionally gotten removed from the default_collectors
class used as the default list for all_collector_classes by
setup.py and compat.py

That made gather_subset['facter'] fail.

mtnbikenc commented:
@alikins Tested again and our openshif_facts.py library now works as expected.

alikins committed:
no TestCase.assertItemsEqual on py3, just use set compares

alikins commented:
@mtnbikenc Excellent. Thanks for testing.
